### ISSUE TITLE
x86 Release Build never reaches prompt on qemu-system-i386 or qemu-system-x86_64

### ISSUE THREAD

**Issue by SilverWingedSeraph** [minoca/os#10]
It's stuck in this state, with or without KVM:
![screenshot from 2016-11-01 07-06-58](https://cloud.githubusercontent.com/assets/3290365/19889242/d9b44890-a001-11e6-92ec-16c43e80100b.png)

**Comment by bonzini** [comment]
I see the same. The EFI build gets to the same point; I also tried AHCI and USB storage to no avail.

The old Windows QEMU build (version 0.13.0) in the starter pack works under Wine though.

**Comment by mellowcandle** [comment]
I'm experiencing the same

**Comment by s-u-p-e-r-t-r-a-m-p** [comment]
could you please how you run qemu? :) I fail due to IO initialization - have just green with some hexa. 
Also is there cmd serial dbg client for linux?

**Comment by bonzini** [comment]
Just `qemu-system-x86_64 pc.img` will do. But I noticed that you're using a very old QEMU. Perhaps you can try with the QEMU images from https://qemu.weilnetz.de/w32/ on Windows?

**Comment by bonzini** [comment]
For serial debug, perhaps you can add `-serial tcp:0.0.0.0:4444,server` (remember punching the port in the firewall!) and connect the server on a Windows machine via TCP.

**Comment by SilverWingedSeraph** [comment]
I can't try on Windows as I don't have a Windows PC available; I'm using the QEMU from the Ubuntu 16.10 repositories, which seems quite old. I'll retry with a newer version.

**Comment by Tester798** [comment]
Same on VMware.

**Comment by evangreen** [comment]
Hey guys. Thanks for the patience, sorry about the bugs.

@SilverWingedSeraph Are you using the binary images from the website, or ones you built yourself?

Sounds like we have trouble running under newer versions of Qemu. I struggled in the past trying to run newer versions of Qemu on Windows, which I believe I debugged down to a livelock issue in Qemu with serial over named pipes. I think I remember it being an issue where we were trying to send serial into Qemu, and the Qemu VM stopped making progress because the named pipe kept polling as ready, but the virtual 16550 wasn't ready, and that loop busy spun. So that's why the Qemu we put in the starter pack is an ancient 0.13.0 build.

I would guess the issue with newer Qemu on Ubuntu is completely different, since the issue I mentioned above is specific to Windows. I'll need to fire up a Ubuntu VM and install Qemu to try and get to the bottom of that one.

@Tester798 What version of VMWare are you using. Can you attach your vmx file?

**Comment by abryrath** [comment]
I am seeing the same problem, running Ubuntu 14.04.05. I have tried with the qemu-system-x86_64 from the Ubuntu repositories (2.0.0) and a source built qemu (2.7.0)

**Comment by Tester798** [comment]
@evangreen Trying on version 12.1.0 with "Other" and "Other 64-bit" profiles with BIOS and EFI boot modes.
On EFI mode after long empty green screen was able too see command prompt but it didn't react to keyboard input.
Attached "Other" BIOS vmx file which gives the result showed on the screenshot from 1st post.
Would be nice to have an option to download ready to use vmx image from the site.
[M.zip](https://github.com/minoca/os/files/564849/M.zip)

**Comment by luislezcair** [comment]
I'm having the same problem, no prompt, like in the first post. I tried all the images built from source (pc.img, pcefi.img and pctiny.img).
qemu 2.7.0 from Arch Linux.

**Comment by ccstevens** [comment]
Thanks for all the input. I've got a repro of this now on Ubuntu 16.10 (64-bit) using QEMU 2.6.1. Oddly, QEMU 2.0.0 worked on 32-bit Ubuntu 14.04.10. 

Will hopefully get to the root cause shortly.

**Comment by evangreen** [comment]
Chris is still working on the issues for Qemu (looks like an ACPI issue). I pushed a minor ACPI change, and I'm now running on VMWare 12.1 and VirtualBox 5.1.8 on my Windows 7 x64 PC. I've added a couple notes about running on VirtualBox to the docs directory.

@Tester798 would you be willing to pull down the latest changes and give VMWare another try?

**Comment by Tester798** [comment]
@evangreen Ok I built the image from source and now in EFI mode I'm getting same result as with BIOS mode (showed in 1st post).
Can you attach your working vmx file?

**Comment by evangreen** [comment]
@Tester798 hmm, so you pulled in the recent ACPI fix and rebuilt? If you want to prove to yourself that your changes are making it into the image you can try messing with the default background color at lib/basevid/textvid.c, line 155. Here's my working vmx configuration, and the vmdk disk images that I handmade. I chose "keep existing format" when attaching the disks.
[vmware.zip](https://github.com/minoca/os/files/569732/vmware.zip)

I was able to boot in both BIOS and EFI mode, though EFI mode was significantly slower to boot.

**Comment by Tester798** [comment]
@evangreen yes, I did git clone and build rel version with toolset.
With your files from vmware.zip I'm getting same result: blinking cursor like in 1st message and after some time cursor stops blinking (but time is still going). For me EFI boots longer too.

Very often problems with emulation for me are because of the AMD cpu I'm using. Maybe current problem is because of this too?
What is your vmware build #? Mine is 3272444. 

Here is the screenshot with changed color:
![image](https://cloud.githubusercontent.com/assets/8844478/19979971/0320daea-a205-11e6-92f8-25c0a7a081ee.png)

**Comment by evangreen** [comment]
@Tester798 Ah, I'm seeing the same thing you are when I try to build the rel version. Can you give it a try with DEBUG=dbg? You can just copy the same tools from x86rel/tools to x86dbg/tools, and remember to use the pc.img at the x86dbg location. If it works with dbg, then we'll at least know I've got a repro of your issue, and it's a temporary workaround for VMWare while I look into the rel failure.

**Comment by Tester798** [comment]
@evangreen Ok I built with DEBUG=dbg and not sure if because of this but the input with keyboard is working now (don't remember if I tested in previous rel build, but it surely didn't work some time ago).
But there is still no command prompt (that's why I didn't test input all the time, thought I should see command prompt).

**Comment by Tester798** [comment]
@evangreen ok I did some testing and this is what I found:

- With images from http://www.minocacorp.com/download/
  - With SATA hdd command prompt doesn't apppear
  - With IDE hdd command prompt is working fine
  - Keyboards input doesn't work

- With custom built images (rel or dbg doesn't matter)
  - Command prompt never appears
  - Keyboard input works

**Comment by evangreen** [comment]
@Tester798 It makes sense that SATA doesn't work, since we don't have an AHCI driver yet. Make sure the VM settings for the custom images are IDE. I'm surprised that dbg doesn't work, maybe there is something to the AMD thing you suggested. I just pushed some fixes for rel builds on my own machine. See if they make any difference, though with your reports of dbg being borked my confidence is lukewarm.

For everyone else, Chris is still working on the Qemu issue.

**Comment by ccstevens** [comment]
The QEMU issue should be fixed if you sync to [785e2c8](https://github.com/minoca/os/commit/785e2c8596a244b074b3b1aec0e60564217b4ac8) and rebuild. I tested it on Ubuntu 16.10 (64-bit) using QEMU 2.6.1.

There ended up being two issues in parsing the ACPI tables. We were not treating stores to the AML constants Zero, One, Ones as no-ops and we were not correctly storing to LocalX Objects. This caused initialization of the PCI routing table to fail.

**Comment by s-u-p-e-r-t-r-a-m-p** [comment]
works well! :) thanks! :)

**Comment by drawkula** [comment]
I still have no prompt with x86dbg and x86rel in Debian-8.6's Qemu-2.1 with `qemu-system-i386` and `qemu-system-x64`.

**Comment by Tester798** [comment]
@evangreen tried again rel build, now it gives no errors while building, but getting same result.
Here is log I'm getting with dbg: http://pastebin.com/kpwK9vU5
Here is 0.2.0.1375 with prompt but not working keyboard: http://pastebin.com/HNLLNGnn
Diff: http://www.diff-online.com/view/581c944c1eb77

For some reason folder /apps/bin is empty in custom built image, but I'm not getting any errors while building it. Here is my build log: http://pastebin.com/nZBdZ6N9

**Comment by heavyrail** [comment]
I've just got latest OS and toolchain from Github, then I've built the OS and did
```qemu-system-x86_64 pc.img```

I've got the following in the terminal
```
WARNING: Image format was not specified for 'pc.img' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
warning: TCG doesn't support requested feature: CPUID.01H:ECX.vmx [bit 5]
```

And I have got green QEMU screen with the single line of text:

```Failure: 0xFFFFFFCC```

What does this mean?

**Comment by drawkula** [comment]
@heavyrail ... instead of `-hda pc.img` try `-drive format=raw,file=pc.img`. This relates to the 1st warning. For the 2nd one I have no idea...

**Comment by heavyrail** [comment]
@drawkula, thank you for your answer. Thanks to you I have found a command which leads to no warnings:
```qemu-system-x86_64 -enable-kvm -cpu host -drive format=raw,file=pc.img```
However, I have still got GSOD (green screen of death) with a single Failure line mentioned above.

Also, I've tried Virtualbox as described in docs/virtualbox.txt. All I have got is the screen with system information but no prompt, similar to the screen posted by topicstarter.

**Comment by ccstevens** [comment]
@s-u-p-e-r-t-r-a-m-p Thanks for sanity checking that the changes don't just work for me.

@drawkula I suspect you're getting more ACPI issues. There is a vast landscape of ACPI tables out there and the various versions of QEMU appear to be a good stress test for our AML interpreter; we'd only tested against the PC's we had lying around and QEMU on Windows. I'll try to get a repro going with QEMU 2.1 and Debian 8.6. Also, given that we've already seen some ACPI issues in the Minoca-pc-1375.zip image, I think we'll have to abandon the pre-built images until we clear these hurdles.

@heavyrail 0xFFFFFFCC is STATUS_CONVERSION_FAILED (-52) and definitely points at a failure in the ACPI AML interpreter. What host system and QEMU version are you running?

**Comment by tejasjadhav** [comment]
@ccstevens @heavyrail Hey same thing happening for me on VirtualBox. Tried with the instructions specified here <https://github.com/minoca/os/blob/master/docs/virtualbox.txt>.

## VirtualBox Configuration:
**RAM:** 256MB
**Chipset:** ICH9 (with I/O APIC enabled)
**Storage:** Generated VMDK file (see the above link for instructions)
**Host:** Ubuntu 16.04 x64

## Screenshot
![virtualbox_minoca_04_11_2016_21_49_17](https://cloud.githubusercontent.com/assets/10093001/20016527/db052b60-a2e5-11e6-96f0-8abf541f0d6e.png)

**Comment by evangreen** [comment]
@Tester798 I've moved our VMWare discussion over to Issue #37 . Let's follow up there.

@heavyrail Not sure I love the coinage of the term GSOD :) It's supposed to be green when it's working too, you know. Of course, given all the troubles I guess no one would know that. Thanks for the patience, all.

**Comment by drawkula** [comment]
@ccstevens ... "jessie-backports" is not black magic, so Qemu-2.7 kind of is available for the current Debian-Stable. But some hardcore stable users reject using the additional backports component of the distribution. It'd be nice when stuff works without it but don't put too much energy into the Qemu-2.1 task.

I got the same empty `bin/apps` direectory issue like mentioned above by @Tester798 on all builds (armv6,armv7,x86)x(dbg,rel). So there are probably 2 different problems with own builds and the Minoca-pc-1375.zip image. It runs with Virtualbox but failed with all Qemus I tested.

**Comment by ccstevens** [comment]
We've got a repro for the empty bin/apps directory. That definitely will cause Qemu to sit without a prompt. @evangreen is tracking a fix for that over on Issue #37 . 

@heavyrail Still has a different issue with the "Failure 0xFFFFFFCC" output. I'm working on a repro of that, but need to find the right version of Qemu.

**Comment by heavyrail** [comment]
@ccstevens,
```
$ qemu-system-x86_64 --version
QEMU emulator version 2.5.0 (Debian 1:2.5+dfsg-5ubuntu10.5), Copyright (c) 2003-2008 Fabrice Bellard
$ uname -a
Linux v580 4.2.0-35-generic #40-Ubuntu SMP Tue Mar 15 22:15:45 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
I'm running Ubuntu 16.04 x86_64 at Lenovo V580 laptop. I use older kernel from Ubuntu 15.10 (4.2.0 instead of 4.4.0) because of troubles with Wi-Fi driver.

**Comment by ccstevens** [comment]
@heavyrail Thanks. Got a repro with Qemu 2.5.0 (still on Ubuntu 16.10).

**Comment by bonzini** [comment]
@evangreen Out of curiosity, why are you not using acpica? As far as I know only Windows (and Minoca OS :)) implemented their own ACPI interpreter from scratch.

**Comment by evangreen** [comment]
Okay, for everybody except heavyrail that is building on Linux, I just pushed change 6d5d50fa3. There was a bug in our build scripts in that the test -nt (newer than) flag fails on Linux if one of the files does not exist. You can verify that you were affected by the problem by checking `$SRCROOT/$ARCH$VARIANT$DEBUG/bin/skel/bin`. If it's absent, or missing sh inside, then this was an issue for you.

Grab the latest, `make clean` in os, and rebuild, and let's see where we are.

**Comment by heavyrail** [comment]
@evangreen, why did you except me from the community of affected? :) I also had empty ```$SRCROOT/$ARCH$VARIANT$DEBUG/bin/skel/bin``` before I update to your latest release and then rebuild. Now the directory mentioned is not empty, but there are still no ```sh``` inside of it. And of course, the OS image still fails to boot either in QEMU or in Virtualbox.

**Comment by drawkula** [comment]
x86rel in Qemu-2.7 (from "jessie-backports") on Debian-8.6:
![20161104-202251-gmt--minoca-x86-rel-in-qemu](https://cloud.githubusercontent.com/assets/9768622/20022975/62f7e7e6-a2dd-11e6-8814-e52ea70c8e0d.png)

I'll downgrade to Qemu-2.1 and update this comment later...

**Comment by evangreen** [comment]
@heavyrail Sorry to single you out, it was just to mention that the 0xFFFFFFCC problem is a different issue that I thought only you were experiencing. Chris is working on that one now.

I had missed one more -nt, pushed 76aaf37fabad593fa93725ba1fc10911cc961d86.

**Comment by evangreen** [comment]
@bonzini Writing it from scratch is kind of my thing, I guess.

**Comment by ccstevens** [comment]
@heavyrail I just pushed a fix. If you sync to [e74c99c](https://github.com/minoca/os/commit/e74c99cf2abbc6abacac432853d736a3f1ca45e4) and rebuild you should get to the command prompt. It was, in fact, another ACPI issue. A DSDT _INI method was depending on a namespace object defined in the SSDT, which we were not loading until after running the _INI method for the DSDT.

@drawkula I've got a repro of the repeated 7's. I'll see what sort of keystrokes we're getting (or missing).

For those now unblocked on QEMU, here's the command I've been using to fire up QEMU. It'll give you networking.

> qemu-system-x86_64 -drive format=raw,file=pc.img -smp 4 -net nic,model=i82559er,macaddr=52:54:00:59:86:03 -net user -m 256

**Comment by heavyrail** [comment]
@evangreen @ccstevens Thank you guys, with https://github.com/minoca/os/commit/2fa9de84a0e1f3b044845b797824329d6cee7856 I can now see the command prompt both in QEMU and Virtualbox.

**Comment by ccstevens** [comment]
A fix for the extra 7's printed when releasing the pointer on QEMU (Ctrl+Alt) is available with [7563dcaf](https://github.com/minoca/os/commit/7563dcafcdf5468dfdcade1dc3ddc01b9a7602b4). QEMU was flushing mouse data to the i8042 device on the Ctrl+Alt. As Minoca doesn't support mice yet, we were interpreting it as keystrokes.

**Comment by drawkula** [comment]
I'm running https://github.com/minoca/os/commit/401ade85071b9cc3c89cccb9d8d368798ae6ed60 now and haven't seen 777777s using it.

**Comment by melezhik** [comment]
confirming, works for me, prompt appears 


```

melezhik@melezhik-pc:~/projects/minoca/os$ qemu-img info /home/melezhik/projects/minoca/x86dbg/bin/pc.img 
image: /home/melezhik/projects/minoca/x86dbg/bin/pc.img
file format: raw
virtual size: 900M (943718400 bytes)
disk size: 15M

melezhik@melezhik-pc:~/projects/minoca/os$ qemu-system-x86_64  -m 2000 -enable-kvm -drive format=raw,file=/home/melezhik/projects/minoca/x86dbg/bin/pc.img

melezhik@melezhik-pc:~/projects/minoca/os$ qemu-system-x86_64 --version
QEMU emulator version 2.0.0 (Debian 2.0.0+dfsg-2ubuntu1.28), Copyright (c) 2003-2008 Fabrice Bellard
melezhik@melezhik-pc:~/projects/minoca/os$ cat /etc/issue
Ubuntu 14.04.5 LTS \n \l

melezhik@melezhik-pc:~/projects/minoca/os$ git log | head -n 4
commit 401ade85071b9cc3c89cccb9d8d368798ae6ed60
Author: Chris Stevens <chris@minocacorp.com>
Date:   Mon Nov 7 17:18:18 2016 -0800

```

**Comment by evangreen** [comment]
Closing the book on this!
